File: Donut/donut.c
#include <stdio.h>
#include <string.h>
#include <math.h>
#include <time.h>   

#define WIDTH 80
#define HEIGHT 22
#define R1 1
#define R2 2
#define K1 150
#define K2 5

int main() {
    float A = 0, B = 0;
    char b[1760]; 
    float z[1760];
    
    printf("\x1b[2J"); 
    
    while (1) {
        memset(b, ' ', sizeof(b));
        memset(z, 0, sizeof(z));
        
        
        for(int k = 0; k < 1760; k++) {
            b[k] = k % 80 == 79 ? '\n' : ' ';
        }
        
        float cA = cos(A), sA = sin(A);
        float cB = cos(B), sB = sin(B);
        
        // Theta loop
        for(float j = 0; j < 6.28; j += 0.07) {
            float ct = cos(j), st = sin(j);
            
            // Phi loop
            for(float i = 0; i < 6.28; i += 0.02) {
                float sp = sin(i), cp = cos(i);
                float h = ct + 2; // R1 + R2*cos(theta)
                float D = 1 / (sp * h * sA + st * cA + 5); // 1/z
                float t = sp * h * cA - st * sA;
                
                int x = 40 + 30 * D * (cp * h * cB - t * sB);
                int y = 12 + 15 * D * (cp * h * sB + t * cB);
                int o = x + 80 * y;
                int N = 8 * ((st * sA - sp * ct * cA) * cB - sp * ct * sA - st * cA - cp * ct * sB);
                
                if(y < 22 && y >= 0 && x >= 0 && x < 79 && D > z[o]) {
                    z[o] = D;
                    b[o] = ".,-~:;=!*#$@"[N > 0 ? N : 0];
                }
            }
        }
        
        printf("\x1b[H"); // Move cursor to home position
        for(int k = 0; k < 1760; k++) {
            putchar(b[k]);
        }
        
        A += 0.07;
        B += 0.03;
        
        // Modern replacement for usleep()
        struct timespec ts;
        ts.tv_sec = 0;
        ts.tv_nsec = 30000 * 1000; 
        nanosleep(&ts, NULL);
    }
    
    return 0;
}

File: Donut/donut.java
import java.util.*;

public class donut {
    public static void main(String[] args) {
        final int WIDTH = 80;
        final int HEIGHT = 22;
        double A = 0, B = 0;
        char[] b = new char[WIDTH * HEIGHT];
        double[] z = new double[WIDTH * HEIGHT];
        
        // Clear screen (works on most terminals)
        System.out.print("\033[2J");
        
        while (true) {
            // Clear buffers
            Arrays.fill(b, ' ');
            Arrays.fill(z, 0);
            
            for (double j = 0; j < 6.28; j += 0.07) {
                double ct = Math.cos(j), st = Math.sin(j);
                for (double i = 0; i < 6.28; i += 0.02) {
                    double sp = Math.sin(i), cp = Math.cos(i);
                    double h = ct + 2;
                    double D = 1 / (sp * h * Math.sin(A) + st * Math.cos(A) + 5);
                    double t = sp * h * Math.cos(A) - st * Math.sin(A);
                    
                    int x = (int)(40 + 30 * D * (cp * h * Math.cos(B) - t * Math.sin(B)));
                    int y = (int)(12 + 15 * D * (cp * h * Math.sin(B) + t * Math.cos(B)));
                    int o = x + WIDTH * y;
                    int N = (int)(8 * ((st * Math.sin(A) - sp * ct * Math.cos(A)) * Math.cos(B) -
                            sp * ct * Math.sin(A) - st * Math.cos(A) - cp * ct * Math.sin(B)));
                    
                    if (y < HEIGHT && y >= 0 && x >= 0 && x < WIDTH && D > z[o]) {
                        z[o] = D;
                        int luminanceIndex = Math.max(0, Math.min(11, N));
                        b[o] = ".,-~:;=!*#$@".charAt(luminanceIndex);
                    }
                }
            }
            
            // Move cursor to home position and print frame
            System.out.print("\033[H");
            for (int j = 0; j < HEIGHT; j++) {
                for (int i = 0; i < WIDTH; i++) {
                    System.out.print(b[i + j * WIDTH]);
                }
                System.out.println();
            }
            
            A += 0.04;
            B += 0.02;
            
            try {
                Thread.sleep(50);
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
                break;
            }
        }
    }
}